### ISSUE TITLE
If using screen reader it's not possible to click the check boxes in the list of tickets to select multiple tickets.

### ISSUE THREAD

**Issue by nktech1135** [freescout-help-desk/freescout#4968]
The subject basically says it all. In a list of tickets i am unable to select multiple tickets using the checkboxes without using the mouse. Typically i should be able to do this with the keyboard if using a screen reader.
System info.
Windows 11, the latest brave and nvda screen reader versions, and freescout 1.8.184
Could someone look into this?

**Comment by freescout-help** [comment]
Do you know which attributes are needed for checkboxes to allow clicking them?

**Comment by nktech1135** [comment]
I don't offhand, i'm not personally a web developer.

**Comment by freescout-help** [comment]
Ok, so we'll leave the issue open. Maybe someone who knows how to fix this will appear.

**Comment by philippkraft** [comment]
This is not a problem that my screen reader using colleague has discovered for herself, although I asked her to check. It might be a problem with this specific screen reader.

**Comment by nktech1135** [comment]
What does she use? i can reproduce this on firefox and brave using the nvda screen reader on win11
I didn't even know there were checkboxes to select multiple tickets until one of the employees pointed it out to me.
Personally, to me it feels like this needs to be fixed in freescout.

**Comment by philippkraft** [comment]
Sorry I can't have a look currently. I think she uses JAWS. I will investigate if it is something she just doesn't know if it exists, or if it is no problem with her screen reader - but that will take some time.

I had a look into the code, the checkbox has a class "magic-checkbox" which uses display:none style on the input itself, I guess to draw the "magic" itself. When I disable the display:none style a second old style checkbox appears under the nice box, which I can access with tabs and select. This might get resolved with either role="checkbox" or aria-hidden="false" as extra attributes where the class magic-checkbox is used. (I have not tested it yet).

**Comment by nktech1135** [comment]
Looks like you're getting somewhere. I'm open to testing this if you can tell me what to change and where.

**Comment by philippkraft** [comment]
You need to edit the file conversations_table.blade.php and change 2 lines in the same way.

These are the two cases in that file:
 
https://github.com/freescout-help-desk/freescout/blob/b0d589c3a4f5bae135957dbb381af63cd56b576e/resources/views/conversations/conversations_table.blade.php#L90

https://github.com/freescout-help-desk/freescout/blob/b0d589c3a4f5bae135957dbb381af63cd56b576e/resources/views/conversations/conversations_table.blade.php#L143

In both lines insert between `class="... magic-checkbox"` and `id="..."`:
`role="checkbox" aria-hidden="false"`

I hope this helps, I am in no way a pro for accessibility. `role` should not be necessary, because the role is already clear by input with the type="checkbox", however the description of aria-hidden sounds fitting: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-hidden, but I am not sure it helps.

**Comment by philippkraft** [comment]
@freescout-help : the main problem here is, that magic-checkbox is not accessible. My (untested) idea above is an ugly work around. Would getting rid of magic-checkbox and use a modern checkbox style eg from bootstrap an alternative?

**Comment by nktech1135** [comment]
I will test @philippkraft's workaround, to see if it works. However, long term it would be good to find a fix that would be acceptable to merge into the project so i don't need to patch each time i upgrade.
From what i know of the bootstrap stuff, i think there code usually works for accessibility so if that would be more acceptable that would work for me.

**Comment by nktech1135** [comment]
I did this and rebooted the server, Still no checkboxes for me, i'm a little puzzled by this.

**Comment by philippkraft** [comment]
Can you post the lines here, so we can be sure everything is correct? I am not sure, if the aria-hidden="false" attribute overwrites the display:none style for the reader, the help texts are a bit unclear about that. It would be better if display:none would not be used, but that is part of the magic-checkbox library used here.

If you also remove the class `magic-checkbox` from these lines, does it work then? (At least we know that it is the magic-checkbox)

**Comment by nktech1135** [comment]
Here you go. Line 90.
            @if (empty($no_checkboxes))<th class="conv-cb"><input type="checkbox" class="toggle-all magic-checkbox" role="checkbox" aria-hidden="false" id="toggle-all"><label for="toggle-all"></label></th>@endif

And line 143
                            <input type="checkbox" class="conv-checkbox magic-checkbox" role="checkbox" aria-hidden="false" id="cb-{{ $conversation->id }}" name="cb_{{ $conversation->id }}" value="{{ $conversation->id }}"><label for="cb-{{ $conversation->id }}"></label>

I'll try removing magic-checkbox to see what happens.

**Comment by nktech1135** [comment]
Removing the magic-checkbox class does work. Is there any way to get a working version of this or something like this back to the project? If i made a pull request, would it be accepted?

**Comment by freescout-help** [comment]
> Removing the magic-checkbox class does work. Is there any way to get a working version of this or something like this back to the project? If i made a pull request, would it be accepted?

Yes, but on regular devices (non screen readers) it should somehow look and work as it works now.

**Comment by philippkraft** [comment]
If the magic-checkbox from some css-library, is there an accessible update? The should not have used display:none, as I read in the last days, but eg. the bootstrap-class ["sr-only"](https://stackoverflow.com/questions/19758598/what-is-sr-only-in-bootstrap-3) or a similar approach (moving the widget with position absolute out of the view port). Or can we patch magic-checkbox?

**Comment by freescout-help** [comment]
> Or can we patch magic-checkbox?

Sure. You can try to replace `display:none` with some class to hide it.

**Comment by nktech1135** [comment]
Somehow it seems to me that if magic-checkbox doesn't work then we should try and use something else.
Or is the issue just that magic checkbox is incorrectly implemented?
I'm lost here with the talk of classes and methods.

**Comment by freescout-help** [comment]
What is the status of this issue?

**Comment by nktech1135** [comment]
I'm waiting on it to be fixed. at this point i'm just not updating my freescout instance.
I'm not a programer so do not know how to add the checkbox classes as you're asking for.
I do think we need to find a solution that works with or without screen readers. I doubt you want to get into detecting if a screen reader is being used or not.
If you give me detailed instructions as given earlier in this thread i'm willing to test things as needed.

**Comment by philippkraft** [comment]
Sorry, I was quite busy with other things. When Iook at the style information of that checkbox with the magic-checkbox class in the browser, it shows display:none as the class's style, defined in a temporary file: https://demo.freescout.net/css/builds/1f34779cd5e6d15a7cde5fb6a5300fbe61518400229.css. The file public/css/magic-check.css is only minified. I am not sure how this is working. After reformatting I get quite in the beginning:

~~~~~~~~~~~~~~.css
.magic-checkbox,
.magic-radio {
    position: absolute;
    display: none
}
~~~~~~~~~~~~~~~~~~~~

I have not tested it, but one way forward could be to incorporate the way Bootstrap 3 uses .sr-only

~~~~~~~~~~~~~~.css
.magic-checkbox,
.magic-radio {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0,0,0,0);
  border: 0;
}
~~~~~~~~~~~~~~~~~

Or just remove ´display:none´ and add the ´sr-only´ class to the two checkboxes. I have currently no test environment for freescout and not the time to set it up. @freescout-help could you test the two versions, if they look ok for you?

On the other hand, the used Bootstrap 3 is outdated anyway, Bootstrap 4 & 5 do have very nice checkbox styles which are screen reader friendly by default. If you are planning to update the Bootstrap version, the problem would be resolved for free and you get rid of the outdated magic-checkbox. My webapp (https://github.com/jlu-ilr-hydro/odmf) took a few hours to upgrade from Bootstrap 4 to Bootstrap 5. Worked very well with regex replace for the most cases.

**Comment by freescout-help** [comment]
Unfortunately we don't have a screen reader to test it. There is no plans to upgrade Bootstrap.

**Comment by nktech1135** [comment]
It's easy to get a screen reader, download nvda from https://nvaccess.org. It's free and open source and is what i use on a daily basis.

**Comment by nktech1135** [comment]
While looking this up this morning i came across the following older issue.
https://github.com/freescout-help-desk/freescout/issues/4412
I wonder if the fix used here would also work for the check boxes in the ticket list?

Anyone more experienced care to test this or tell me what to change on my instance to do so?

**Comment by philippkraft** [comment]
I have patched my solution in our instance - it looks exactly as before. I have also checked the solution for #4412 (which is similar but simpler) and it works also nicely on the display. Since the latter is easier and already used in d41fac7a7f406432fadae7fd7103659dcf2367ff, I am now going for that solution.

I wrote my blind colleague an Email, when she responds and it works, I can create a PR.

**Comment by nktech1135** [comment]
ok, thanks to claude, i got this working. see pull request https://github.com/freescout-help-desk/freescout/pull/5274
